src/toil/job.py

```diff
@@ -1470,6 +1540,39 @@ def run(self, fileStore):
         rValue = userFunction(*((self,) + tuple(self._args)), **self._kwargs)
         return rValue
 
+
+class PromisedRequirementFunctionWrappingJob(FunctionWrappingJob):
+    """
+    Creates a child function using parent function parameters and promised resource requirements.
+
+    (see :class:`toil.job.FunctionWrappingJob` and :class:`toil.job.Job`)
+    """
+
+    def run(self, fileStore):
+        userFunction = self._getUserFunction()
+        self._kwargs["cores"] = self.cores_promise() if hasattr(self, "cores_promise") else self.cores
+        self._kwargs["disk"] = self.disk_promise() if hasattr(self, "disk_promise") else self.disk
+        self._kwargs["memory"] = self.memory_promise() if hasattr(self, "memory_promise") else self.memory
+        self._kwargs["cache"] = self.cache_promise() if hasattr(self, "cache_promise") else self.cache
+        return self.addChildFn(userFunction, *self._args, **self._kwargs).rv()
+
+
+class PromisedRequirementJobFunctionWrappingJob(FunctionWrappingJob):
+    """
+    Creates a child job function using parent function parameters and promised resource requirements.
+
+    (see :class:`toil.job.JobFunctionWrappingJob and :class:`toil.job.Job`)
+    """
+
+    def run(self, fileStore):
+        userFunction = self._getUserFunction()
+        self._kwargs["cores"] = self.cores_promise() if hasattr(self, "cores_promise") else self.cores
```

Comment:
hasattr(self, ...) is a bad smell. Attributes should always be initialized in the constructor. Assigning None is ok and `is None` can be used to test wether they are set.